components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/util/ClientCredentialAuthConfig.java

```diff
@@ -0,0 +1,392 @@
+     *
+     * @return Access token
+     * @throws FrameworkException {@link FrameworkException}
+     */
+    private void validateRequiredProperties(Map<String, Object> properties) throws FrameworkException {
+
+        if (!properties.containsKey(CONSUMER_KEY_VARIABLE_NAME) ||
+                !properties.containsKey(CONSUMER_SECRET_VARIABLE_NAME) ||
+                !properties.containsKey(TOKEN_ENDPOINT)) {
+            asyncReturn.accept(authenticationContext, Collections.emptyMap(), OUTCOME_FAIL);
+            LOG.error("Required properties not defined. Aborting token request.");
+            throw new FrameworkException("Missing required properties.");
+        }
+    }
+
+    /**
+     * This method is used to get the access token from the cache or request a new token from the token endpoint.
+     *
+     * @return Access token
+     * @throws FrameworkException {@link FrameworkException}
+     */
+    private String getAccessToken() throws FrameworkException {
+        String accessToken = apiAccessTokenCache.getValueFromCache(getConsumerKey(),
+                authenticationContext.getTenantDomain());
+        try {
+            if (StringUtils.isNotEmpty(accessToken) && !isTokenExpired(accessToken)) {
+                LOG.debug("Unexpired access token available in cache.");
+                return accessToken;
+            } else {
+                // Attempt the first request for an access token
+                LOG.info("Attempting initial access token request for session data key: " +
+                        authenticationContext.getContextIdentifier());
+                accessToken = requestAccessToken(); // Attempt to request a new token
+                if (accessToken != null) {
+                    return accessToken;
+                }
+                if (LoggerUtils.isDiagnosticLogsEnabled()) {
+                    DiagnosticLog.DiagnosticLogBuilder diagnosticLogBuilder = new
+                            DiagnosticLog.DiagnosticLogBuilder(Constants.LogConstants.ADAPTIVE_AUTH_SERVICE,
+                            Constants.LogConstants.ActionIDs.RECEIVE_TOKEN);
+                    diagnosticLogBuilder.inputParam(Constants.LogConstants.InputKeys.TOKEN_ENDPOINT, getTokenEndpoint())
+                            .configParam(Constants.LogConstants.ConfigKeys.SUPPORTED_GRANT_TYPES,
+                                    GRANT_TYPE_CLIENT_CREDENTIALS)
+                            .resultMessage("Initial request failed, proceeding with retry attempts.")
+                            .logDetailLevel(DiagnosticLog.LogDetailLevel.APPLICATION)
+                            .resultStatus(DiagnosticLog.ResultStatus.FAILED);
+                    LoggerUtils.triggerDiagnosticLogEvent(diagnosticLogBuilder);
+                }
+                // If the initial request fails, proceed with retry logic
+                LOG.info("Initial request failed, proceeding with retry attempts.");
```

Comment:
Better move this log to accessTokenRequestRetry

Reply:
Resolved with https://github.com/wso2-extensions/identity-conditional-auth-functions/pull/157/commits/01b5e2f5f849d7a4e89d792e28845597544fca39